semgrep-core/bin/Main.ml

```diff
@@ -264,6 +275,55 @@ let gen_layer ~root ~query file =
   ()
 (*e: function [[Main_semgrep_core.gen_layer]] *)
 
+let filemtime file =
+  (Unix.stat file).Unix.st_mtime
+
+(* The function below is mostly a copy-paste of Common.cache_computation.
+ * This function is slightly more flexible because we can put the cache file
+ * anywhere thanks to the argument 'cache_file_of_file'.
+ * We also try to be a bit more type-safe by using the version tag above.
+ * TODO: merge in pfff/commons/Common.ml at some point
+ *)
+let cache_computation file cache_file_of_file f =
+  if not !use_parsing_cache
+  then f ()
+  else begin
+    if not (Sys.file_exists file)
+    then begin
+      pr2 ("WARNING: cache_computation: can't find file "  ^ file);
+      pr2 ("defaulting to calling the function");
+      f ()
+    end else begin
+    profile_code "Main.cache_computation" (fun () ->
+
+      let file_cache = cache_file_of_file file in
+      if Sys.file_exists file_cache && filemtime file_cache >= filemtime file
+      then begin
+        if !verbose then pr2 ("using cache: " ^ file_cache);
+        let (version, res) = Common2.get_value file_cache in
+        if version != ast_version
+        then failwith (spf "Version mismatch! Clean the cache file %s"
+                      file_cache);
+        res
+      end
+      else begin
+        let res = f () in
+        Common2.write_value (ast_version, res) file_cache;
+        res
+      end
+      )
+    end
+  end
+
+
+let cache_file_of_file filename =
+  let dir = spf "/tmp/semgrep_core_cache_%d" (Unix.getuid()) in
+  if not (Sys.file_exists dir)
+  then Unix.mkdir dir 0o700;
+  (* hopefully there will be no collision *)
+  let md5 = Digest.string filename in
```

Comment:
What's the purpose of hashing the filename here?

Reply:
Just to create easily a filename in /tmp/semgrep_core_cache<uid>/ that correspond to a complex path. semgrep will pass me  /path/to/project/nested/dir/foo.c  and I will hash that.
I could also replace all / with __, but I might get escaping issues, so just using the md5sum of the path seemed simpler. Also it prevents people to play with the cache directory. Those filename are a bit more opaque than a /tmp/semgrep_core/path__to__project__nested__dir__foo_c

Reply:
Gotcha. What does happen if there's a collision? Can we at least store the full file path in the cache data, along with the `NamedAST`? That way we can tell when we've hit a collision.

Reply:
Separate conversation about `/tmp`: you can use `Filename.get_temp_dir_name ()` instead of assuming `/tmp` exists. (this point was raised by @mschwager )

from https://caml.inria.fr/pub/docs/manual-ocaml/libref/Filename.html :
```
val get_temp_dir_name : unit -> string
```
> The name of the temporary directory: Under Unix, the value of the TMPDIR environment variable, or "/tmp" if the variable is not set. Under Windows, the value of the TEMP environment variable, or "." if the variable is not set. The temporary directory can be changed with Filename.set_temp_dir_name.
>
> Since 4.00.0
```

Reply:
Good idea!